compiler/rustc_builtin_macros/src/autodiff.rs

```diff
@@ -0,0 +1,707 @@
+                ItemKind::Fn(box ast::Fn { sig, .. }) => sig,
+                _ => {
+                    ecx.sess
+                        .dcx()
+                        .emit_err(errors::AutoDiffInvalidApplication { span: item.span() });
+                    return vec![item];
+                }
+            };
+            (sig.clone(), false)
+        }
+        Annotatable::AssocItem(ref assoc_item, _) => {
+            let sig = match &assoc_item.kind {
+                ast::AssocItemKind::Fn(box ast::Fn { sig, .. }) => sig,
+                _ => {
+                    ecx.sess
+                        .dcx()
+                        .emit_err(errors::AutoDiffInvalidApplication { span: item.span() });
+                    return vec![item];
+                }
+            };
+            (sig.clone(), true)
+        }
+        _ => {
+            ecx.sess.dcx().emit_err(errors::AutoDiffInvalidApplication { span: item.span() });
+            return vec![item];
+        }
+    };
+
+    let meta_item_vec: ThinVec<NestedMetaItem> = match meta_item.kind {
+        ast::MetaItemKind::List(ref vec) => vec.clone(),
+        _ => {
+            ecx.sess.dcx().emit_err(errors::AutoDiffInvalidApplication { span: item.span() });
+            return vec![item];
+        }
+    };
+
+    let has_ret = sig.decl.output.has_ret();
+    let sig_span = ecx.with_call_site_ctxt(sig.span);
+
+    let (vis, primal) = match &item {
+        Annotatable::Item(ref iitem) => (iitem.vis.clone(), iitem.ident.clone()),
+        Annotatable::AssocItem(ref assoc_item, _) => {
+            (assoc_item.vis.clone(), assoc_item.ident.clone())
+        }
+        _ => {
+            ecx.sess.dcx().emit_err(errors::AutoDiffInvalidApplication { span: item.span() });
+            return vec![item];
+        }
+    };
+
+    // create TokenStream from vec elemtents:
+    // meta_item doesn't have a .tokens field
+    let comma: Token = Token::new(TokenKind::Comma, Span::default());
+    let mut ts: Vec<TokenTree> = vec![];
+    for t in meta_item_vec.clone()[1..].iter() {
+        let val = first_ident(t);
+        let t = Token::from_ast_ident(val);
+        ts.push(TokenTree::Token(t, Spacing::Joint));
+        ts.push(TokenTree::Token(comma.clone(), Spacing::Alone));
+    }
+    if !sig.decl.output.has_ret() {
+        // We don't want users to provide a return activity if the function doesn't return anything.
+        // For simplicity, we just add a dummy token to the end of the list.
+        let t = Token::new(TokenKind::Ident(sym::None, false.into()), Span::default());
+        ts.push(TokenTree::Token(t, Spacing::Joint));
+    }
+    let ts: TokenStream = TokenStream::from_iter(ts);
+
+    let x: AutoDiffAttrs = from_ast(ecx, &meta_item_vec, has_ret);
+    if !x.is_active() {
+        // We encountered an error, so we return the original item.
+        // This allows us to potentially parse other attributes.
+        return vec![item];
+    }
```

Comment:
If we encounter an error, we set the mode inactive. In that case our macro doesn't expand to anything, we don't generate the placeholder function (thus prevent our enzyme backend from crashing on an invalid combination) and just give a nice high level rust error.

Reply:
This seems a bit strange, I wonder if it makes sense to have a `DiffMode::Err` which is explicitly for error-handling, instead of inactive (which I assume could be due to non-error causes)?

Reply:
I renamed it from inactive to Error.